Exile Implant Revamp: Borders are Closed

GreatBigFailure opened the pull request:
tl;dr: If you implant someone it won't do anything until they first leave the Z they were implanted on. Then they can't return to it.

I’ve reworked exile implants to be more useful, more flexible, and I think more fun. 

**Why**
The old exile implants just bound you to specifically Z5. This meant you could only explore the roid and on roid station they were flat useless. They also didn’t actually un-exile you when you had the implant removed. This is a good idea but poor execution for a non-lethal way to handle traitors that isn’t perma. We need more of those so here’s another. 

**What**

Exile implants now record the Z you were implanted on. When implanting someone you can also add a disable phrase. The implant doesn’t do anything until they’ve first left the Z level. Once their Z has changed they’ll receive a message whenever re-entering the banned Z. If they don’t leave within 10 seconds they’ll be randomly teleported to a tile within any Z besides centcomm and their banned Z.

The disable phrase or EMPs jaunts the implant back to the tile you were first implanted. So it can work as a notifier that it was removed. EMPs aren’t guaranteed however and can cause the crude wormhole tech in the implant to malfunction.

This is a basic version since it’s already a fairly big addition and could impact players as is. I've added exile implants to cargo as well. I have plans for different versions but for now let's try this out.

:cl:
 * rscadd: Exile implants now work properly as a non-lethal prisoner removal method.

SonixApache commented:
Be *sure* to test this with regards to nuke disk fuckery.

Dacendeth commented:
wont the random teleport sometimes teleport them inside the roid and they'd just be stuck?

GreatBigFailure commented:
> Be _sure_ to test this with regards to nuke disk fuckery.

I knew I forgot to test something. I'll do that in a few minutes.

cloutcode commented:
>not delivering a lethal dose of neurotoxin to whoever dares to try border hopping for their sins

GreatBigFailure commented:
> wont the random teleport sometimes teleport them inside the roid and they'd just be stuck?

I don't know how that never happened while I was testing but I 100% do not have any safeties in place for that. I'll add some, shouldn't be hard.

DrSnips commented:
I don't think any changes to gateway is needed. With antags they are if anything showing you mercy by letting you drift in space rather than just killing you. 200 is also a little steep for the basic level of implants if you ask me.

> Exile implants now record the Z you were implanted on

Usually this is going to be Z1. Does this mean Z1 is the one they are locked TO or locked FROM?

GreatBigFailure commented:
My mindset with the gateway is that since this is a revamp and not some new exciting item people likely aren't going to think to seek it out. Seeing it on the cargo list might change that. Exile implants also aren't even associated with the gateway anymore so it's just weird that they're in there. 
200 was a spitball number but you're right, a bit lower at least. Much cheaper if people want me to leave it how it is, a fair bit more expensive if they want the bluespace crystal thing or similar. 

> Does this mean Z is the one they are locked TO or locked FROM?
Locked FROM. 
Once you leave the Z you were implanted on the implant won't let you return to it. So you could have a Trader ban a crook from their base or have a Xenoarch ban a miner who destroyed all their digs from the roid. I doubt either of those will ever happen but it's now flexible is the point.

GreatBigFailure commented:
There, can't be teleported into a wall anymore. 

I also tested the disk and it's fine. It just prevents you from teleporting. Any other interaction with the disk would either require it to not be on station Z or just be a normal man trying to carry it off the Z and failing. Tried both teleporting out of Z1 by being banned and teleporting to other Z levels with EMP malfunction.

Viciousarus commented:
>Teleporting them
>Not gibbing them
Weak implant

Dacendeth commented:
Wont someone be able to return if they just wear a tinfoil hat, as it prevents teleporting?

GreatBigFailure commented:
> Wont someone be able to return if they just wear a tinfoil hat, as it prevents teleporting?

Yep, but they'll get stunned every 10 seconds and start receiving minor damage slowly. Each teledeport has a 50 50 chance of causing 5 clone and 5 brute. I'm thinking of removing that though.

TGFH commented:
Why you would remove the implants from the gateway if the goal is to get people to actually use the implants? I get that the gateway thing is defunct but walking to the gateway for free implants > going to cargo and paying + waiting for the cargonians to send the crate to the wrong department.

GreatBigFailure commented:
Mostly it just bugs me that they're there, I admit. I can ask a mapper to just put them in sec instead then add them to cargo as well? Basically how loyalty implants currently work. Gateway is such a goofy place for them.

tuypo1 commented:
> not delivering a lethal dose of neurotoxin to whoever dares to try border hopping for their sins

Problem with that is you would be floating through space and if you happened to hit the station z you would just die

GreatBigFailure commented:
>not delivering a lethal dose of neurotoxin to whoever dares to try border hopping for their sins

>Weak implant

The issue, like Tuypo said is that you don't control what Z level you enter when transitioning. There's a bunch of different factors that might see them accidentally return to Z1. I also wanted to make it flexible since that's what makes the best stories. This opens possibilities for dumb stuff like a teleport wizard that exiles himself then teleports in, bullshits around for 10 seconds, then gets auto-teleported away. 
Besides, it's meant to be non-lethal.

SonixApache commented:
would be a lot better if people just couldn't drift onto their banned z-level in the first place

GreatBigFailure commented:
> would be a lot better if people just couldn't drift onto their banned z-level in the first place

Why do you say that? 
The teleport is for flexibility and **_emergent gameplay_**. It's also to make it hard to get rid of but not necessarily a 100% permanent timeout chair from video games. I'm open to being convinced though.

SonixApache commented:
- people will get 'punished' for something they cannot really prevent - drifting onto the main Z
- people will actively try to launch themselves onto the banned z-level to teleport, hoping they'll hit something nice like a vault
 - people can and will get stuck in extremely gay places such as roid tunnels

I'm cool with the teleport functionality, it applies well to when people hop onto shuttles or find some other fucked up way to get back onto their banned Z-level.
But really, they should just be unable to get there in the first place in as many ways as possible. Subconscious evasion, or something.

GreatBigFailure commented:
The teleport isn't much of a punishment, that's one reason I'm thinking I should get rid of the clone damage. 
That is a good point about actively trying to get to the banned Z for intentional warping. Originally that's why I added the clone damage but now I'm torn. 
The tunnels are a problem I hadn't considered. Fair point.
The mass driver and gateway are incorrect though. They don't teleport away until they've first left the banned Z once. They're fair game aside from that. 
>Teleported to anywhere in an away mission
Wait what. I thought away missions were vaults spawned on z2?

I'll look into how I would go about what you've suggested. I can see a lot of reason for it but it also reduces the flexibility a bit so I'm on the fence.

SonixApache commented:
Away missions are spawned on Z-6.
Good point on the mass driver, but my gateway issue still stands; you can throw them through the hoop, but they can then just walk back through the hoop on their end and end up somewhere else.

GreatBigFailure commented:
>Away missions are spawned on Z-6.

I might need to just make it only teleport them to space turfs then. I guess I could do an area check but I've also realized they have a very, very small chance of just getting trapped forever in a vault secret room or something. That would also fix the mining tunnel issue. 
Away missions might have teleport prevention? I assume they do but I haven't checked. Either way I might add the space turf thing because it would fix so many potential issues, including most you've presented, in one clean move.

GreatBigFailure commented:
Exile implants will now only teleport you to space tiles, they can still teleport you right next to a vault but not into one, an away mission, or a mining tunnel. 

I also added exile lockboxes to cargo. They're 150 credits but a lockbox has 3 implants in it. Gateway implants were replaced by 2 lockboxes, same amount of implants but it's at least less weird now. Lockbox access is armory.

Finally I changed the brute/clone damage from a 50% chance to cause 5 of each to a 100% chance to cause up to 5 of each, randomly.

DamianX commented on code/game/objects/items/weapons/implants/exile2.dm:
```diff
@@ -1,39 +1,125 @@
-* Instead of confining someone within an away mission, this locks someone to the asteroid.
-*/
+
+//There now it's fun
```
```suggestion
```

DamianX commented on code/game/objects/items/weapons/implants/exile2.dm:
```diff
@@ -1,39 +1,125 @@
 /obj/item/weapon/implant/exile
 	name = "exile"
-	desc = "Prevents you from returning from the asteroid"
+	desc = "Prevents returning to where you were implanted"
```
```suggestion
	desc = "Prevents returning to where you were implanted."
```

DamianX commented on code/game/objects/items/weapons/implants/exile2.dm:
```diff
@@ -1,39 +1,125 @@
-	desc = "Prevents you from returning from the asteroid"
+	desc = "Prevents returning to where you were implanted"
+	var/illegalZ = null
+	var/mob/living/theExile = null
```
implants already have a var to track the mob they're implanted into, called `imp_in`, which is a terrible name but it's probably better than using a special snowflake var for the same purpose.

GreatBigFailure replied:
I can't change that, the implant requires typecasting to work. Won't even register as implanted in the person if I use imp_in. Even the previous version of the implant had a typecast variable to replace imp_in.

DamianX commented on code/game/objects/items/weapons/implants/exile2.dm:
```diff
@@ -1,39 +1,125 @@
-	I.locked_to_z = ASTEROID_Z
+/obj/item/weapon/implant/exile/implanted(mob/source as mob)
+	theExile = source
+	disablePhrase = input("Choose phrase to disable:") as text
```
`input`s without a cancel option are annoying.
calling `input` without passing a mob will default to usr, which is bad.
use this instead:
```suggestion
	disablePhrase = stripped_input(user, "Choose phrase to disable:")
```
The phrasing is a bit awkward as well.
And you should now handle the case in which disablePhrase ends up being null, and `return 0` accordingly

DamianX commented on code/game/objects/items/weapons/implants/exile2.dm:
```diff
@@ -1,39 +1,125 @@
-	var/mob/living/carbon/I = M
-	to_chat(I, "<span class='notice'>You shiver as you feel a weak bluespace void surround you.</span>")
-	I.locked_to_z = ASTEROID_Z
+/obj/item/weapon/implant/exile/implanted(mob/source as mob)
```
It looks like the proc definition `/obj/item/weapon/implant/proc/implanted(var/mob/source)` only supports a single parameter but in truth callers of this proc pass a second argument. See implanter.dm line 33. All definitions and calls of this proc should be adjusted to reflect this, like so:
```suggestion
/obj/item/weapon/implant/exile/implanted(mob/source, mob/user)
```

DamianX commented on code/game/objects/items/weapons/implants/exile2.dm:
```diff
@@ -1,39 +1,125 @@
+/obj/item/weapon/implant/exile/emp_act()
+	if(malfunction)
+		return
+	malfunction = 1
```
pick one:
- consistently set this to fixed values like here
- consistently decrease/increase this value, like you did in the spawn()

DamianX commented on code/game/objects/items/weapons/implants/exile2.dm:
```diff
@@ -1,39 +1,125 @@
+	#define FREEDOM 1
+	#define RANDOM_TELEPORT 2
+	#define IMPLANTED_SITE_PORT 3
+	switch(pick(1,2,3))
```
```suggestion
	switch(pick(FREEDOM,RANDOM_TELEPORT,IMPLANTED_SITE_PORT))
```

DamianX commented on code/game/objects/items/weapons/implants/exile2.dm:
```diff
@@ -1,39 +1,125 @@
+		return
+	malfunction = 1
+	var/mob/living/M = imp_in
+	#define FREEDOM 1
```
`#undef` these once you don't need them anymore

DamianX commented on code/game/objects/items/weapons/implants/exile2.dm:
```diff
@@ -1,39 +1,125 @@
 
+/obj/item/weapon/implant/exile/proc/zBan()
+	var/turf/T = get_turf(src)
+	switch(beenSpaced)
```
this switch is awkward, just use `if()` and `else()`

DamianX commented on code/game/objects/items/weapons/implants/exile2.dm:
```diff
@@ -1,39 +1,125 @@
+		var/turf/T = get_turf(src)
+		if(T.z == illegalZ)
+			var/warpZ = pick(zlevels)
+			var/warpTo = locate(rand(5,world.maxx - 10), rand(5, world.maxy - 10), warpZ)
```
`TRANSITIONEDGE` is how far away from the map border one teleports to a different level, you could use it instead of magic numbers

DamianX commented:
> The disable phrase or EMPs jaunts the implant back to the tile you were first implanted.

I don't understand why you would do this.
It doesn't make sense to me that the implant would teleport back to the original location.
I think upon hearing the specified phrase the implant should just stop functioning, and/or be removed.

GreatBigFailure commented:
>It doesn't make sense to me that the implant would teleport back to the original location.
It's so that, if you put the extra effort in to implant them somewhere you go a lot, you can see it there and know they've disabled it. It's so that exiling an antag doesn't then give them the element of surprise. It makes them more tempting/easy to use. 
It also helps explain the EMP effect of jaunting them back to their implant location. I wanted that because it will give them 10 seconds of panic/opportunity and just be funny.

SonixApache commented:
you gotta remove the file from the rsc if it's no longer being loaded
open up DM, should do it
